### PR TITLE
feat(transport):add grpc and http with listen

### PR DESCRIPTION
**Pull request by haiyux**
#### Which issue(s) this PR fixes (resolves / be part of):

fixes #

### CODE REVIEW

File: transport/grpc/server.go [hunk 1]
```diff
@@ -150,21 +156,6 @@ func NewServer(opts ...ServerOption) *Server {
 // examples:
 //   grpc://127.0.0.1:9000?isSecure=false
 func (s *Server) Endpoint() (*url.URL, error) {
-	s.once.Do(func() {
-		lis, err := net.Listen(s.network, s.address)
```
Comment (Casper-Mars) [inline transport/grpc/server.go:154]: would be it satisfy if just replace this line by s.lis(if s.lis is not nil)?

Comment (haiyux) [reply transport/grpc/server.go:154]: Yes，I want to get rid of the side effects of endpoint。

**Comment by letian0805** [comment]
Hey! This is the third PR about the side effects of Listener and Endpoint, right?
The first one is https://github.com/go-kratos/kratos/pull/1470

**Comment by haiyux** [comment]
> Hey! This is the third PR about the side effects of Listener and Endpoint, right? The first one is #1470

Your PR does not eliminate the side effects of endpoint, it still listen in the endpoint function.

**Comment by letian0805** [comment]
> > Hey! This is the third PR about the side effects of Listener and Endpoint, right? The first one is #1470
> 
> Your PR does not eliminate the side effects of endpoint, it still listen in the endpoint function.

Yes i know, so longxboy created another one to fix side effects. https://github.com/go-kratos/kratos/pull/1497

**Comment by haiyux** [comment]
feat #1714

**Review by tonybase — APPROVED** [review]

**Pull request merged**